src/ripple/ledger/impl/ApplyStateTable.cpp

```diff
@@ -0,0 +1,589 @@
+    {
+    case Action::erase:
+        return false;
+    case Action::cache:
+    case Action::insert:
+    case Action::modify:
+        break;
+    }
+    if (! k.check(*sle))
+        return false;
+    return true;
+}
+
+auto
+ApplyStateTable::succ (ReadView const& base,
+    key_type const& key, boost::optional<
+        key_type> const& last) const ->
+            boost::optional<key_type>
+{
+    boost::optional<key_type> next = key;
+    items_t::const_iterator iter;
+    // Find base successor that is
+    // not also deleted in our list
+    do
+    {
+        next = base.succ(*next, last);
+        if (! next)
+            break;
+        iter = items_.find(*next);
+    }
+    while (iter != items_.end() &&
+        iter->second.first == Action::erase);
+    // Find non-deleted successor in our list
+    for (iter = items_.upper_bound(key);
+        iter != items_.end (); ++iter)
+    {
+        if (iter->second.first != Action::erase)
+        {
+            // Found both, return the lower key
+            if (! next || next > iter->first)
+                next = iter->first;
+            break;
+        }
+    }
+    // Nothing in our list, return
+    // what we got from the parent.
+    if (last && next >= last)
+        return boost::none;
+    return next;
+}
+
+std::shared_ptr<SLE const>
+ApplyStateTable::read (ReadView const& base,
+    Keylet const& k) const
+{
+    auto const iter = items_.find(k.key);
+    if (iter == items_.end())
+        return base.read(k);
+    auto const& item = iter->second;
+    auto const& sle = item.second;
+    switch (item.first)
+    {
+    case Action::erase:
+        return nullptr;
+    case Action::cache:
+    case Action::insert:
+    case Action::modify:
+        break;
+    };
+    if (! k.check(*sle))
+        return nullptr;
+    return sle;
+}
+
+std::shared_ptr<SLE>
+ApplyStateTable::peek (ReadView const& base,
+    Keylet const& k)
+{
+    auto iter = items_.lower_bound(k.key);
+    if (iter == items_.end() ||
+        iter->first != k.key)
+    {
+        auto const sle = base.read(k);
+        if (! sle)
+            return nullptr;
+        // Make our own copy
+        using namespace std;
+        iter = items_.emplace_hint (iter,
+            piecewise_construct,
+                forward_as_tuple(sle->key()),
+                    forward_as_tuple(Action::cache,
+                        make_shared<SLE>(*sle)));
+        return iter->second.second;
+    }
+    auto const& item = iter->second;
+    auto const& sle = item.second;
+    switch (item.first)
+    {
+    case Action::erase:
+        return nullptr;
+    case Action::cache:
+    case Action::insert:
+    case Action::modify:
+        break;
+    };
+    if (! k.check(*sle))
+        return nullptr;
+    return sle;
+}
+
+void
+ApplyStateTable::erase(
+    ReadView const& base,
+        std::shared_ptr<SLE> const& sle)
+{
+    auto const iter =
+        items_.find(sle->key());
+    if (iter == items_.end())
```

Comment:
OK.  So I think none of these should be a logic error.  Instead, I think we should log a warning.

In my reading, we should only call `LogicError` when we are _forced_ to do so.  In other words, we don't do it if we suspect something is wrong - we should only terminate _if there is no way we can continue in this code path_.

In this method, this is not the case.  Someone asked to erase something that doesn't exist, or even something with the wrong key.

Well, we succeeded.  The item does not exist in the ledger.  Perhaps we suspect something is wrong - but this code path can continue fine.